openhcl/openhcl_boot/src/arch/x86_64/tdx.rs

```diff
@@ -94,3 +96,31 @@ impl minimal_rt::arch::IoAccess for TdxIoAccess {
         let _ = tdcall::tdcall_io_out(&mut TdcallInstruction, port, data as u32, 1);
     }
 }
+
+/// Reads MSR using TDCALL
+pub fn read_msr_tdcall(msr_index: u32) -> u64 {
+    let mut msr_value: u64 = 0;
+    tdcall_rdmsr(&mut TdcallInstruction, msr_index, &mut msr_value).unwrap();
+    msr_value
+}
+
+/// Global variable to store tsc frequency.
+static mut TSC_FREQUENCY: u64 = 0;
+
+/// Gets the timer ref time in 100ns, and 0 if it fails to get it
+pub fn get_tdx_tsc_reftime() -> u64 {
+    // SAFETY: no concurrent accessors. Only BSP running shim_main calls it first.
+    unsafe {
+        if TSC_FREQUENCY == 0 {
+            TSC_FREQUENCY = read_msr_tdcall(HV_X64_MSR_TSC_FREQUENCY);
```

Comment:
Do we trust the hypervisor for this value, because it's coming via the host? Or what happens if the hypervisor returns something invalid?

Reply:
Do we need to tag this as a follow up to figure out?

Reply:
Yes, Another way to get the tsc frequency is using CPUID 0x15. I will add a comment to explore this.